constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ConstraintLayoutStatistics.java

```diff
@@ -0,0 +1,264 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.constraintlayout.widget;
+
+import android.annotation.SuppressLint;
+import android.util.Log;
+
+import androidx.constraintlayout.core.Metrics;
+
+import java.text.DecimalFormat;
+import java.util.ArrayList;
+
+/**
+ * This provide metrics of the complexity of the layout that is being solved.
+ * The intent is for developers using the too to track the evolution of their UI
+ * Typically the developer will monitor the computations on every callback of
+ * mConstraintLayout.addOnLayoutChangeListener(this::callback);
+ *
+ */
+public class ConstraintLayoutStatistics {
+    public final static int NUMBER_OF_LAYOUTS = 1;
+    public final static int NUMBER_OF_ON_MEASURES= 2;
+    public final static int NUMBER_OF_CHILD_VIEWS= 3;
+    public final static int NUMBER_OF_CHILD_MEASURES= 4;
+    public final static int DURATION_OF_CHILD_MEASURES= 5;
+    public final static int DURATION_OF_MEASURES = 6;
+    public final static int DURATION_OF_LAYOUT = 7;
+    public final static int NUMBER_OF_VARIABLES = 8;
+    public final static int NUMBER_OF_EQUATIONS = 9;
+    public final static int NUMBER_OF_SIMPLE_EQUATIONS = 10;
+
+    private final Metrics mMetrics = new Metrics();
+    ConstraintLayout mConstraintLayout;
+    private static int MAX_WORD = 25;
+    private static final String WORD_PAD = new String(new char[MAX_WORD]).replace('\0', ' ');
+
+    /**
+     * Measure performance information about ConstraintLayout
+     *
+     * @param constraintLayout
+     */
+    public ConstraintLayoutStatistics(ConstraintLayout constraintLayout) {
+        attach(constraintLayout);
+    }
+
+    /**
+     * Copy a layout Stats useful for comparing two stats
+     * @param copy
+     */
+    public ConstraintLayoutStatistics(ConstraintLayoutStatistics copy) {
+        mMetrics.copy(copy.mMetrics);
+    }
+
+
+    /**
+     * Attach to a ConstraintLayout to gather statistics on its layout performance
+     * @param constraintLayout
+     */
+    public void attach(ConstraintLayout constraintLayout) {
+        constraintLayout.fillMetrics(mMetrics);
+        mConstraintLayout = constraintLayout;
+    }
+
+    /**
+     * Detach from a ConstraintLayout
+     */
+    public void detach() {
+        if (mConstraintLayout != null) {
+            mConstraintLayout.fillMetrics(null);
+        }
+    }
+
+    /**
+     * Clear the current metrics
+     */
+    public void reset() {
+        mMetrics.reset();
+    }
+
+    /**
+     * Create a copy of the statistics
+     * @return a copy
+     */
+    public ConstraintLayoutStatistics clone() {
+        return new ConstraintLayoutStatistics(this);
+    }
+
+    private String fmt(DecimalFormat df, float val, int sp) {
```

Comment:
Maybe we don't need a JavaDoc comment for private method, but it would be helpful if we could have a simple one line comment for what this function for (e.g, format xxx)? Same as the functions below. However, it's totally optional

Reply:
done